### DISCUSSION TITLE
Porting code with `Nullable<>` to 2.0.0

### DISCUSSION THREAD

**Discussion by ensc** [diesel-rs/diesel#2766] [Q&A] [answered]
I am about to port my existing code to work with the master branch and stumble about problems with `Nullable<>`.  The following code compiles fine with 1.4.6 but fails on master:

```rust
use diesel::dsl::{Or, Eq};
use diesel::expression::bound::Bound;
use diesel::expression::{ Expression };
use diesel::sql_types::{ Bool, Text, Nullable };

type FiBound<FI> = Bound<<FI as Expression>::SqlType, &'static str>;
type BoolBound = Bound<Bool, bool>;

fn filter<FI>(field: FI) -> Or<BoolBound,Eq<FI, FiBound<FI>>>
where
    FI: diesel::expression_methods::ExpressionMethods<SqlType = Nullable<Text>>
{
    use diesel::expression::AsExpression;
    use diesel::BoolExpressionMethods;

    let fn_bool = |v| <bool as AsExpression<Bool>>::as_expression(v);

    fn_bool(true).or(field.eq(""))
}

fn main() {
}
```

Error on master is

```
error[E0271]: type mismatch resolving `<Grouped<diesel::expression::operators::Eq<FI, diesel::expression::bound::Bound<diesel::sql_types::Nullable<Text>, &'static str>>> as diesel::Expression>::SqlType == Bool`
  --> src/main.rs:9:1
   |
9  | / fn filter<FI>(field: FI) -> Or<BoolBound,Eq<FI, FiBound<FI>>>
10 | | where
11 | |     FI: diesel::expression_methods::ExpressionMethods<SqlType = Nullable<Text>>
12 | | {
...  |
18 | |     fn_bool(true).or(field.eq(""))
19 | | }
   | |_^ expected struct `diesel::sql_types::Nullable`, found struct `Bool`
   |
   = note: expected struct `diesel::sql_types::Nullable<Bool>`
              found struct `Bool`
   = note: required because of the requirements on the impl of `AsExpression<Bool>` for `Grouped<diesel::expression::operators::Eq<FI, diesel::expression::bound::Bound<diesel::sql_types::Nullable<Text>, &'static str>>>`
```

Replacing the `Nullable<Text>` with `Text` allows it to build (but I need `Nullable` here).

How can I express this with master?

For reference: real code at
- https://gitlab-ext.sigma-chemnitz.de/ensc/sstate-server/-/blob/8b492651d2b867502f23bcaefe233f0144f4f615/src/models/lookup/string.rs#L49-74
- https://gitlab-ext.sigma-chemnitz.de/ensc/sstate-server/-/blob/8b492651d2b867502f23bcaefe233f0144f4f615/src/models/user.rs#L412-415

**Comment by weiznich** [comment, marked as answer]
We changed our type tracking to be more in line with what database actually do. `NULL OR some_expr` evaluates to `NULL`, therefore we made sure that boolean expressions propagate nullability based on the arguments.
In consequence that means as soon as at least on of your arguments may be nullable diesel requires that the output can be nullable. Due to type inference issues this can require a third type parameter to `diesel::dsl::Or`, as outlined [in the documentation](https://docs.diesel.rs/master/diesel/expression_methods/trait.BoolExpressionMethods.html#method.or), to specify if that expression is `Nullable` or not.

```rust
use diesel::dsl::{Or, Eq, AsExprOf};
use diesel::sql_types::{ Bool, Text, Nullable };
use diesel::{IntoSql, ExpressionMethods};

type FiBound<ST> = AsExprOf<&'static str, ST>;
type BoolBound = AsExprOf<bool, Bool>;

fn filter<FI>(field: FI) -> Or<BoolBound, Eq<FI, FiBound<FI::SqlType>>, Nullable<Bool>>
where
    FI: ExpressionMethods<SqlType = Nullable<Text>>,
{
    use diesel::BoolExpressionMethods;

    true.into_sql::<Bool>().or(field.eq(""))
}
```

(I also removed any usage of API's that are not part of our public API.)
